qiskit/pulse/schedule.py

```diff
@@ -438,6 +515,88 @@ def _add_timeslots(self, time: int, schedule: ScheduleComponent) -> None:
                         "".format(new=schedule.name or '', old=self.name or '', time=time,
                                   ch=channel, t0=interval[0], tf=interval[1]))
 
+        _check_nonnegative_timeslot(self._timeslots)
+
+    def replace(self,
+                old: ScheduleComponent,
+                new: ScheduleComponent,
+                inplace: bool = False,
+                ) -> 'Schedule':
+        """Return a schedule with the ``old`` instruction replaced with a ``new``
+        instruction.
+
+        The replacment matching is based on an instruction equality check.
+
+        .. jupyter-kernel:: python3
+          :id: replace
+
+        .. jupyter-execute::
+
+          from qiskit import pulse
+
+          d0 = pulse.DriveChannel(0)
+
+          sched = pulse.Schedule()
+
+          old = pulse.Play(pulse.Constant(100, 1.0), d0)
+          new = pulse.Play(pulse.Constant(100, 0.1), d0)
+
+          sched += old
+
+          sched = sched.replace(old, new)
+
+          assert sched == pulse.Schedule(new)
+
+        Only matches at the top-level of the schedule tree. If you wish to
+        perform this replacement over all instructions in the schedule tree.
+        Flatten the schedule prior to running::
+
+        .. jupyter-execute::
+
+          sched = pulse.Schedule()
+
+          sched += pulse.Schedule(old)
+
+          sched = sched.flatten()
+
+          sched = sched.replace(old, new)
+
+          assert sched == pulse.Schedule(new)
+
+        Args:
+          old: Instruction to replace.
+          new: Instruction to replace with.
+          inplace: Replace instruction by mutably modifying this ``Schedule``.
+
+        Returns:
+          The modified schedule with ``old`` replaced by ``new``.
+
+        Raises:
+            PulseError: If the ``Schedule`` after replacements will has a timing overlap.
+        """
+        new_children = []
+        for time, child in self._children:
+            if child == old:
+                new_children.append((time, new))
+            else:
+                new_children.append((time, child))
+
+        try:
+            new_sched = Schedule(*new_children)
+        except PulseError as err:
+            raise PulseError(
+                'Replacement of {old} with {new} results in '
+                'overlapping timeslots.'.format(
+                    old=old, new=new)) from err
+
+        if inplace:
+            self._duration = 0
+            self._timeslots = new_sched.timeslots
+            self.__children = new_sched.__children
+            return self
```

Comment:
I don't understand how this works. Can you explain it to me?

I see that it's tested below (`sched.replace(old, new, inplace=True)`) Can you add a check there that `sched.duration` is correct after, and not `0`? It also looks like some of the timeslots would be removed by these lines; can you check those, too?

Also, why make a new schedule if `inplace` is true?

Reply:
Need to make a new Schedule, because it is only possible to add_timeslots and not remove. I can try adding remove.